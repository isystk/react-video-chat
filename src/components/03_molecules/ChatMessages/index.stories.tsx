import { Meta, Story } from '@storybook/react'
import React from 'react'
import ChatMessages from './index'
import MainService from '@/services/main'

export default {
  title: '03_molecules/ChatMessages',
  component: ChatMessages,
} as Meta

const main = {
  self: {
    connectionId: 'aaa',
    name: 'aaa',
    photo: 'images/friends/BigBoss.png',
  },
  members: {
    aaa: {
      connectionId: 'aaa',
      name: 'aaa',
      photo: 'images/friends/BigBoss.png',
    },
  },
  chanels: {
    all: {
      id: 'all',
      name: 'すべて',
      type: 'all',
      photo: 'images/friends/Alpha_Team.png',
      detail: 'ルーム内のすべてのメンバー',
      chat: {
        chanelId: 'test',
        messages: [
          {
            type: 'text',
            data: 'hello',
            sendId: 'aaa',
            datetime: Date(),
          },
        ],
      },
    },
  },
  selectChanelId: 'all',
} as MainService

const Template: Story = () => <ChatMessages main={main} />

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
