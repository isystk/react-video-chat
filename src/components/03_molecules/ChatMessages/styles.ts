import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles((theme) => ({
  chatMessage: {
    padding: '20px 20px 0 20px',
    backgroundColor: '#f1f1f1',
    '& .message_row': {
      display: 'grid',
      gridTemplateColumns: '70%',
      marginBottom: '20px',
    },
    '& .message-text': {
      padding: '9px 14px',
      fontSize: '16px',
      marginBottom: '5px',
    },
    '& .message-time': {
      fontSize: '12px',
      color: '#777',
    },
    '& .other-message .message-time': {
      width: '66px',
      marginLeft: '65px',
    },
    '& .you-message': {
      justifyContent: 'end',
      '& .message-content': {
        justifyItems: 'end',
        display: 'grid',
      },
      '& .message-text': {
        color: 'rgb(73, 73, 73)',
        backgroundColor: 'rgb(92, 238, 92)',
        border: '1px solid rgb(92, 238, 92)',
        borderRadius: '14px 14px 0 14px',
      },
    },
    '& .other-message': {
      justifyItems: 'start',
      '& .message-content': {
        gridTemplateColumns: '48px 1fr',
        gridColumnGap: '15px',
        display: 'grid',
      },
      '& .message-text': {
        color: '#111',
        backgroundColor: '#eee',
        border: '1px solid #ddd',
        borderRadius: '14px 14px 14px 0',
      },
    },
    '& .head': {
      width: '50px',
      height: '50px',
      borderRadius: '50%',
    },
  },
}))

export { useStyles }
