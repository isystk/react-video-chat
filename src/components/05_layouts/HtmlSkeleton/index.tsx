import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Header from '@/components/04_organisms/Header'
import SideMenu from '@/components/04_organisms/SideMenu'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { CssBaseline } from '@material-ui/core'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  main,
  title,
  description,
  isMenuOpen,
  setMenuOpen,
  children,
  ...props
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
    </Head>
    <Context.Provider value={main}>
      <CssBaseline />
      <Header isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
      <div>{children}</div>
      <SideMenu isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
    </Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()
  const [isMenuOpen, setMenuOpen] = useState(false)

  if (!main) return <></>

  const APP_NAME = process.env.APP_NAME

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : React.cloneElement(child)
  )
  if (!title) title = process.env.APP_NAME

  const description = `${APP_NAME} is an application that allows you to casually chat with your friends!`
  return presenter({
    main,
    title,
    description,
    isMenuOpen,
    setMenuOpen,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
