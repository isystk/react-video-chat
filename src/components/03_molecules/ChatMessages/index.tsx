import * as React from 'react'
import { Stamps } from '@/services/Chat'
import { FC, useContext, useEffect, useState } from 'react'
import * as _ from 'lodash'
import moment from 'moment'
import { ContainerProps, WithChildren } from 'types'
import { useStyles } from './styles'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

/** ChatMessagesProps Props */
export type ChatMessagesProps = WithChildren
/** Presenter Props */
export type PresenterProps = ChatMessagesProps & {
  main
  classes
  windowHeight
  appStyle
}

/** Presenter Component */
const ChatMessagesPresenter: FC<PresenterProps> = ({
  main,
  classes,
  windowHeight,
  appStyle,
  ...props
}) => (
  <>
    <div id="chatRoom" className="chat_message" style={appStyle(windowHeight)}>
      {main.chanels[main.selectChanelId].chat.messages.map((message, index) => {
        const isMe = message.sendId === main.self.connectionId
        console.log(message.sendId, main.self.connectionId)
        const member = main.members[message.sendId]
        const isStamp = 'stamp' === message.type
        return (
          <div
            key={index}
            className={`message_row ${isMe ? 'you-message' : 'other-message'}`}
          >
            <div className="message-content">
              {!isMe && <img className="head" src={member.photo} alt="" />}
              {isStamp ? (
                <img className="head" src={Stamps[message.data]} alt="" />
              ) : (
                <div className="message-text">{message.data}</div>
              )}
              <div className="message-time">
                {moment(message.datetime).format('MM/DD HH:mm')}
              </div>
            </div>
          </div>
        )
      })}
    </div>
  </>
)

/** Container Component */
const ChatMessagesContainer: React.FC<
  ContainerProps<ChatMessagesProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>
  const classes = useStyles()
  const [windowHeight, setWindowHeight] = useState(0)

  useEffect(() => {
    setWindowHeight(window.innerHeight)
  }, [])

  if (_.size(main.chanels) === 0) return <></>

  const appStyle = (vh) => {
    return {
      height: vh - 64 - 69,
      overflowY: 'scroll',
    }
  }

  return presenter({
    children,
    main,
    classes,
    windowHeight,
    appStyle,
    ...props,
  })
}

export default connect<ChatMessagesProps, PresenterProps>(
  'ChatMessages',
  ChatMessagesPresenter,
  ChatMessagesContainer
)
