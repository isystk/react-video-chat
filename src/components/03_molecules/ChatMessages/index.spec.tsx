import React from 'react'
import renderer from 'react-test-renderer'
import ChatMessages from './index'
import '@testing-library/jest-dom/extend-expect'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'
import ChanelService from '@/services/chanel'

describe('ChatMessages', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))
    main.setName('isystk')
    main.room.rooms = { '1': { id: '1', name: 'test', description: 'test' } }
    main.room.setRoomId('1')
    main.addChanel(
      new ChanelService(
        main,
        'all',
        'すべて',
        'all',
        'images/friends/Alpha_Team.png',
        'ルーム内のすべてのメンバー'
      )
    )
    main.setChanelId('all')
    main.addMember({
      connectionId: 'bbb',
      name: 'bbb',
      photo: 'images/friends/David.png',
    })
    main.self.connectionId = 'aaa'
    main.chanels['all'].chat.messages = [
      {
        type: 'text',
        data: 'hello',
        chanelId: 'all',
        sendId: 'aaa',
        datetime: '2022-09-07 12:00:00',
      },
      {
        type: 'stamp',
        data: 'smile',
        chanelId: 'all',
        sendId: 'aaa',
        datetime: '2022-09-07 13:00:00',
      },
      {
        type: 'text',
        data: 'see you',
        chanelId: 'all',
        sendId: 'bbb',
        datetime: '2022-09-07 14:00:00',
      },
    ]

    const component = renderer.create(
      <Context.Provider value={main}>
        <ChatMessages />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
