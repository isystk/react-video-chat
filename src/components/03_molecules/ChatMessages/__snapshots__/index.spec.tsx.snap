// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessages Match Snapshot 1`] = `
<div
  className="makeStyles-chatMessage-1"
  style={
    Object {
      "height": -133,
      "overflowY": "scroll",
    }
  }
>
  <div
    className="message_row you-message"
  >
    <div
      className="message-content"
    >
      <div
        className="message-text"
      >
        hello
      </div>
      <div
        className="message-user"
      />
      <div
        className="message-time"
      >
        9/7 12:00
      </div>
    </div>
  </div>
  <div
    className="message_row you-message"
  >
    <div
      className="message-content"
    >
      <img
        alt=""
        className="head"
        src="images/emoji/smile.png"
      />
      <div
        className="message-user"
      />
      <div
        className="message-time"
      >
        9/7 13:00
      </div>
    </div>
  </div>
  <div
    className="message_row other-message"
  >
    <div
      className="message-content"
    >
      <img
        alt=""
        className="head"
        src="images/friends/David.png"
      />
      <div
        className="message-text"
      >
        see you
      </div>
      <div
        className="message-user"
      >
        bbb
      </div>
      <div
        className="message-time"
      >
        9/7 14:00
      </div>
    </div>
  </div>
</div>
`;
